### PR TITLE
ASoC: rt712-sdca: add the function for version B

### PR DESCRIPTION
**Pull request by shumingfan**
The version B will support the multi-lane function and integrate the DMIC function in one SoundWire interface.
Due to some registers having different default values between version A and B, this patch also removes the redundant default registers to avoid confusion.

### CODE REVIEW

File: sound/soc/codecs/rt712-sdca.c [hunk 1]
```diff
@@ -763,21 +771,21 @@ static int rt712_sdca_fu05_event(struct snd_soc_dapm_widget *w,
...
 	case SND_SOC_DAPM_PRE_PMD:
 		regmap_write(rt712->regmap,
 			SDW_SDCA_CTL(FUNC_NUM_JACK_CODEC, RT712_SDCA_ENT_USER_FU05,
-				RT712_SDCA_CTL_FU_MUTE, CH_L),
+				RT712_SDCA_CTL_FU_MUTE, CH_01),
 				mute);
 		regmap_write(rt712->regmap,
 			SDW_SDCA_CTL(FUNC_NUM_JACK_CODEC, RT712_SDCA_ENT_USER_FU05,
-				RT712_SDCA_CTL_FU_MUTE, CH_R),
+				RT712_SDCA_CTL_FU_MUTE, CH_02),
```
Comment (plbossart) [inline sound/soc/codecs/rt712-sdca.c:788]: it might be simpler to review is the CH_L/CH_01 and CH_R/CH_02 were moved to an initial patch without any change of functionality?

File: sound/soc/codecs/rt712-sdca.h [hunk 2]
```diff
@@ -207,6 +238,11 @@ enum rt712_sdca_hw_id {
 
 #define RT712_PART_ID_713 0x713
 
+enum rt712_sdca_version {
+	RT712_VA,
+	RT712_VB,
+};
+
```
Comment (plbossart) [inline sound/soc/codecs/rt712-sdca.h:251]: is there anything that differentiates the VA and VB at the device_ID level?
Also is there a way to know if the dmic is supported or not based on ACPI information?
not sure how we would change the machine driver to add the dmic path or not?

@bardliao what do you think?

File: sound/soc/codecs/rt712-sdca.h [hunk 3]
```diff
@@ -191,6 +221,7 @@ struct  rt712_sdca_priv {
 enum {
 	RT712_AIF1,
 	RT712_AIF2,
+	RT712_AIF3,
```
Comment (plbossart) [inline sound/soc/codecs/rt712-sdca.h:230]: how can we make sure that AIF3 is never used with a VA chip?

I find the whole thing a bit complicated if the information is hidden the machine integration could be problematic, no?

also how would we select a topology that sometimes has a dmic dailink and sometimes not?

**Review by plbossart — COMMENTED** [review]
Thanks for sharing @shumingfan, I have rather big concerns on the topology/machine driver support if the VA/VB information is not visible at the ACPI level.

File: sound/soc/codecs/rt712-sdca.h (hunk 2 above)
Comment (bardliao) [reply sound/soc/codecs/rt712-sdca.h:251]: We could use `dai_info->quirk` to make the DMIC be optional, like https://github.com/thesofproject/linux/pull/4848/commits/47bae326da237ca6cf0414dc5a8a823d1ff83a39

File: sound/soc/codecs/rt712-sdca.h (hunk 3 above)
Comment (bardliao) [reply sound/soc/codecs/rt712-sdca.h:230]: It is better to have different ADR between -VA and -VB. Or we might construct topology name dynamically, something like https://github.com/thesofproject/linux/pull/4847/commits/0930b9424667422db3fc7695bb525e89fa530e9d. But to be honest, I think it will be painful to do it.

File: sound/soc/codecs/rt712-sdca.c (hunk 1 above)
Comment (shumingfan) [reply sound/soc/codecs/rt712-sdca.c:788]: Thanks for reviewing. I will make an initial patch for this change.

File: sound/soc/codecs/rt712-sdca.h (hunk 2 above)
Comment (plbossart) [reply sound/soc/codecs/rt712-sdca.h:251]: yes but that quirk is defined at the machine driver level. It's too late to select the topology, which is entirely based on ACPI information.

**Comment by plbossart** [comment]
@shumingfan is the function listed in the DSDT? IIRC all the functions should be a "Device" under the codec device, so in theory we might be able to add a quirk to go read the function and decide which system it is.

IOW, we would first start from the ACPI _ADR, but add a quirk to go check if there is a subdevice representing the DMIC function. And in that case we would know for sure what the codec is and select topology.

**Comment by shumingfan** [comment]
> IOW, we would first start from the ACPI _ADR, but add a quirk to go check if there is a subdevice representing the DMIC function. And in that case we would know for sure what the codec is and select topology.

@plbossart Regarding RT712VA, we might check ACPI _ADR that the part id 171X exists or not.
Regarding RT712VB or VA, could we check the DisCo property "mipi-sdca-terminal-type"?
If mipi-sdca-terminal-type belongs to 0x201 or 0x205, that machine will use the DMIC function.
image

**Comment by plbossart** [comment]
> @plbossart Regarding RT712VA, we might check ACPI _ADR that the part id 171X exists or not. 

Yes that's a possibility.

>Regarding RT712VB or VA, could we check the DisCo property "mipi-sdca-terminal-type"? If mipi-sdca-terminal-type belongs to 0x201 or 0x205, that machine will use the DMIC function.

It's probably better to detect the function type, we should see a SmartMic function.

I'll start cooking something.

**Comment by plbossart** [comment]
@shumingfan see https://github.com/thesofproject/linux/pull/4930 for an example on how we can parse the DSDT information and figure out if there are any SDCA functions in a device. we could use this to check if the device supports the  SmartMic function.

Of course that assumes ACPI information is valid, which is already not true on the first device I tried...

**Comment by shumingfan** [comment]
> Of course that assumes ACPI information is valid, which is already not true on the first device I tried...

@plbossart  Thanks. The PR#4930 looks good to me. The function type of DMIC of RT713 might be a simple mic function.
image

**Comment by plbossart** [comment]
The SimpleMic topology is not yet in the SDCA spec, it's still being added for 1.1. So I don't know how if it's really what it is, and how one would infer the topology mask. 

That's my main fear with this ACPI-based solution, it's a mix of older devices with an older unpublished SDCA/DisCo spec, not sure how we can deal with all the variants. It might be just easier to only support 1.0+...

File: sound/soc/codecs/rt712-sdca-sdw.c [hunk 4]
```diff
@@ -178,13 +194,15 @@ static int rt712_sdca_read_prop(struct sdw_slave *slave)
 	unsigned long addr;
 	struct sdw_dpn_prop *dpn;
 
+	sdw_slave_read_prop(slave);
+
```
Comment (plbossart) [inline sound/soc/codecs/rt712-sdca-sdw.c:198]: how is this related to the VA/VB support?

File: sound/soc/codecs/rt712-sdca-sdw.c [hunk 5]
```diff
@@ -178,13 +194,15 @@ static int rt712_sdca_read_prop(struct sdw_slave *slave)
...
 	prop->scp_int1_mask = SDW_SCP_INT1_BUS_CLASH | SDW_SCP_INT1_PARITY;
 	prop->quirks = SDW_SLAVE_QUIRKS_INVALID_INITIAL_PARITY;
 
 	prop->paging_support = true;
 
 	/* first we need to allocate memory for set bits in port lists */
-	prop->source_ports = BIT(4); /* BITMAP: 00010000 */
+	prop->source_ports = BIT(8) | BIT(4); /* BITMAP: 100010000 */
```
Comment (plbossart) [inline sound/soc/codecs/rt712-sdca-sdw.c:205]: I am really confused, is the port8 supported on all versions?

File: sound/soc/codecs/rt712-sdca.c [hunk 6]
```diff
@@ -1031,6 +1432,10 @@ static int rt712_sdca_pcm_hw_params(struct snd_pcm_substream *substream,
 	if (!rt712->slave)
 		return -EINVAL;
 
+	/* VA doesn't support AIF3 */
+	if (dai->id == RT712_AIF3 && rt712->version_id == RT712_VA)
+		return -EINVAL;
```
Comment (plbossart) [inline sound/soc/codecs/rt712-sdca.c:1437]: in what cases would it be necessary to the AIF3?

I am really concerned that the machine driver would need to know about VA/VB.

It's really problematic if the same device has different capabilities over time. It's no longer the same device IMHO and the partID should have been modified.

**Review by plbossart — CHANGES_REQUESTED** [review]
@shumingfan I really don't know how we can support the VB at the machine driver level, this is really really problematic. The opens raised in the initial review have not been solved....

File: sound/soc/codecs/rt712-sdca-sdw.c (hunk 4 above)
Comment (shumingfan) [reply sound/soc/codecs/rt712-sdca-sdw.c:198]: @plbossart This is related to the VB capability for multi-lanes. 
I tested with the PR https://github.com/thesofproject/linux/pull/4733.
I think there is no effect on the VA.

File: sound/soc/codecs/rt712-sdca-sdw.c (hunk 5 above)
Comment (shumingfan) [reply sound/soc/codecs/rt712-sdca-sdw.c:205]: The port8 supports on the VB. 
The DMIC function will use port8 and the driver only creates the "rt712-sdca-aif3" of DMIC DAI when using the VB.
It is easy to add port8 here and allocate memory for set bits in port lists.
The VA could not use port8 because it would not create the DMIC DAI.

File: sound/soc/codecs/rt712-sdca.c (hunk 6 above)
Comment (shumingfan) [reply sound/soc/codecs/rt712-sdca.c:1437]: AIF3 is the DMIC function. The VB integrates the DMIC function in one soundwire interface which supports the multi-lanes. The customers could use the DMIC of RT712 if they don't want another DMIC codec.

> It's really problematic if the same device has different capabilities over time. It's no longer the same device IMHO and the partID should have been modified.

Hmm...I'm sorry for that. I can't decide the partID if the codec will be modified.

**Comment by plbossart** [comment]
@shumingfan for my education, what's the difference between RT712 VB and RT722? They both seem to support UAJ, SmartAmp and SmartMic functions, along with multi-lane.

**Comment by shumingfan** [comment]
> @shumingfan for my education, what's the difference between RT712 VB and RT722? They both seem to support UAJ, SmartAmp and SmartMic functions, along with multi-lane.

@plbossart As far as I know, RT712VB has a newer SDCA version than RT722.
RT712VB has another function like TV mode, however, it isn't implemented in the codec driver yet.
The basic functions like UAJ/Amp/Mic are similar to RT722.

File: sound/soc/codecs/rt712-sdca.c [hunk 7]
```diff
@@ -1307,6 +1716,131 @@ int rt712_sdca_io_init(struct device *dev, struct sdw_slave *slave)
...
+	pm_runtime_get_noresume(&slave->dev);
+
+	rt712_sdca_index_read(rt712, RT712_VENDOR_REG, RT712_JD_PRODUCT_NUM, &val);
+	rt712->hw_id = (val & 0xf000) >> 12;
+	rt712->version_id = (val & 0x0f00) >> 8;
+	dev_dbg(&slave->dev, "%s hw_id=0x%x, version_id=0x%x\n", __func__, rt712->hw_id, rt712->version_id);
+
+	if (rt712->version_id == RT712_VA)
+		rt712_sdca_va_io_init(rt712);
+	else {
```
Comment (plbossart) [inline sound/soc/codecs/rt712-sdca.c:1834]: I would suggest adding a test here

````c
if (!sdca_device_quirk_match(slave, SDCA_QUIRKS_RT712_VB)) {
    dev_err(&slave->dev, "%s: RT712 VB detected with no SmartMic Function in ACPI, invalid configuration\n", __func__);
    return -EINVAL;
}
````

that way we will not have a case where we trip between hardware and ACPI config.

**Review by plbossart — COMMENTED** [review]
@shumingfan I suggest rebasing on https://github.com/thesofproject/linux/pull/5010 to make sure ACPI and hardware are aligned.

**Comment by shumingfan** [comment]
The patches were upstreamed already.
